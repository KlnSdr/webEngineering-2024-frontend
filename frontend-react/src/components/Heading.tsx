import React from "react";

interface HeadingProps {
  headingText: string;
}

const Heading: React.FC<HeadingProps> = ({ headingText }) => {
  return <h1 className="headline-with-lines">{headingText}</h1>;
};

export default Heading;
