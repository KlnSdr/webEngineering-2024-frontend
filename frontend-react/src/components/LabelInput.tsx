import React, { useState, useEffect } from "react";
import Heading from "./Heading";

/**
 * Props for the LabelInput component.
 * @typedef {Object} LabelInputProps
 * @property {string} labelText - The text to display as the label.
 * @property {string} initialValue - The initial value of the input field.
 * @property {function} onChange - Callback function to handle changes in the input value.
 */
interface LabelInputProps {
  labelText: string;
  initialValue: string;
  onChange: (value: string) => void;
}

/**
 * LabelInput component renders a labeled input field with a heading.
 * @param {LabelInputProps} props - The props for the component.
 * @returns {JSX.Element} The rendered LabelInput component.
 */
const LabelInput: React.FC<LabelInputProps> = ({
  labelText,
  initialValue,
  onChange,
}) => {
  const [value, setValue] = useState(initialValue);

  useEffect(() => {
    setValue(initialValue);
  }, [initialValue]);

  return (
    <div>
      <Heading headingText={labelText} />
      <input
        type="text"
        value={value}
        onChange={(e) => {
          setValue(e.target.value);
          onChange(e.target.value);
        }}
        className="form-control"
      />
    </div>
  );
};

export default LabelInput;