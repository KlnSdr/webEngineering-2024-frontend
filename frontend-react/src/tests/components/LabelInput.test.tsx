import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import '@testing-library/jest-dom/extend-expect';
import LabelInput from "../../components/LabelInput";
import Heading from "../../components/Heading";

jest.mock("../../components/Heading", () => ({ headingText }: { headingText: string }) => <div>{headingText}</div>);

describe("LabelInput Component", () => {
    test("renders with initial value and label text", () => {
        const labelText = "Test Label";
        const initialValue = "Initial Text";
        const handleChange = jest.fn();

        render(<LabelInput labelText={labelText} initialValue={initialValue} onChange={handleChange} />);

        const heading = screen.getByText(labelText);
        const input = screen.getByRole("textbox");

        expect(heading).toBeInTheDocument();
        expect(input).toHaveValue(initialValue);
    });

    test("calls onChange callback when value changes", () => {
        const labelText = "Test Label";
        const initialValue = "";
        const handleChange = jest.fn();

        render(<LabelInput labelText={labelText} initialValue={initialValue} onChange={handleChange} />);

        const input = screen.getByRole("textbox");
        userEvent.type(input, "New Value");

        expect(handleChange).toHaveBeenCalledTimes(9); // "New Value" is 9 characters long
        expect(handleChange).toHaveBeenCalledWith("N");
        expect(handleChange).toHaveBeenCalledWith("Ne");
        expect(handleChange).toHaveBeenCalledWith("New");
        expect(handleChange).toHaveBeenCalledWith("New ");
        expect(handleChange).toHaveBeenCalledWith("New V");
        expect(handleChange).toHaveBeenCalledWith("New Va");
        expect(handleChange).toHaveBeenCalledWith("New Val");
        expect(handleChange).toHaveBeenCalledWith("New Valu");
        expect(handleChange).toHaveBeenCalledWith("New Value");
    });

    test("updates value when initialValue prop changes", () => {
        const labelText = "Test Label";
        const initialValue = "Initial Text";
        const newValue = "New Initial Text";
        const handleChange = jest.fn();

        const { rerender } = render(<LabelInput labelText={labelText} initialValue={initialValue} onChange={handleChange} />);

        let input = screen.getByRole("textbox");
        expect(input).toHaveValue(initialValue);

        rerender(<LabelInput labelText={labelText} initialValue={newValue} onChange={handleChange} />);
        input = screen.getByRole("textbox");
        expect(input).toHaveValue(newValue);
    });
});
