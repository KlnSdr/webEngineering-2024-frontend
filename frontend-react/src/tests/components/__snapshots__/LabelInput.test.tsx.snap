// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelInput Component renders with initial value and label text 1`] = `
<DocumentFragment>
  <div>
    <div>
      Test Label
    </div>
    <input
      class="form-control"
      type="text"
      value="Initial Text"
    />
  </div>
</DocumentFragment>
`;
